### ISSUE TITLE
type of output values not printed anymore

### ISSUE THREAD

**Issue by nthiery** [QuantStack/xeus-cling#192]
In earlier versions of xeus-cling, the type of the value of a cell was printed out:
```
     In [1]   1 + 1
     Out[1]  (int) 2
```
While preparing the exercise sheets for a class tomorrow, I realized early this morning that this is no more the case:
```
     In [1]   1 + 1
     Out[1]  2
```

We used this feature quite extensively for teaching, because it gave more insight to the student; as a side effect, it also made a distinction between outputs (with cout) and cell values (with return).

I can see the point in other use cases of *not* having the type. There are also cases where the type infos is large and not that usable due to template expansion.

Is this something that could be made configurable one way or the other? If at all possible with the current version of xeus-cling?

```
xeus                      0.14.1                  cling_0  [cling]  QuantStack
xeus-cling                0.4.7                         0    QuantStack
```

**Comment by SylvainCorlay** [comment]
Indeed, we don't print the type anymore. (This is a consequence of the work that was done with respect to the rich mime type rendering.)

Labelling as a feature request.

**Comment by nthiery** [comment]
>    Indeed, we don't print the type anymore. (This is a consequence of the
>    work that was done with respect to the rich mime type rendering.)
> 
>    Labelling as a feature request.

Ok. Thanks for the quick feedback. You don't have a quick an dirty way
to activate it on an existing installation? possibly recompiling some
code if needed?

**Comment by JohanMabille** [comment]
@nthiery unfortunately we don't have a quick and dirty way to fix that, it requires a backward incompatible change that would break the rich display mechanism for third party libraries. So we need to release a new version.

**Comment by SylvainCorlay** [comment]
Hum, I think that would mostly be trouble (to downgrade).

A *very* ugly hack workaround would be to modify the default implementation of `mime_bundle_repr` in https://github.com/QuantStack/xeus-cling/blob/master/include/xcpp/xmime.hpp#L22 to use prepend the return value with `typeid(value).name()`.

This should not require a rebuild because that function is interpreted at runtime by cling.

**Comment by nthiery** [comment]
>    A very ugly hack workaround would be to modify the default
>    implementation of mime_bundle_repr in
>    [1]https://github.com/QuantStack/xeus-cling/blob/master/include/xcpp/xm
>    ime.hpp#L22 to use prepend the return value with typeid(value).name().

Yeah! That works like a charm :-)

Sylvain you saved my day; thanks so much. Even better, I could
customize the output further to make it super explicit to best match
the teaching needs:

      > 1 + 1
      2
      type: int

It's fun that the interpreted nature of cling brings us all the
flexibility that we have in Python :-)

Besides it's not that ugly actually; in fact pretty close from a clean
solution: replacing that function with a function pointer (or
something of that nature) would make for a natural hook/call-back that
could be customized by applications.

**Comment by SylvainCorlay** [comment]
The issue with that solution is that it may be returning mangled names depending on the platform.

We have a cleaner solution but unfortunately, it would be a more significant change.

**Comment by nthiery** [comment]
>    The issue with that solution is that it may be returning mangled names
>    depending on the platform.

I indeed had to demangle here, but the following trick (from
stackoverflow?) worked; I have no idea how portable it is though:

```
template<class C>
char * type(C c) {
    int status;
    char * demangled;
    return abi::__cxa_demangle(typeid(c).name(),0,0,&status);
}
```

**Comment by SylvainCorlay** [comment]
Long story short,  we have the real type name handy before the call to mime bundle repr and we should pass it as a second argument to the function. But this would be backward incompatible...
